File: bench/testdata/samples/A10/PSB-A10-PREV-001/python/vulnerable.py
# PSB-A10-PREV-001: URL preview/unfurl feature
# CWE: CWE-918
# Expected: GTSS-SSRF-001
from flask import Flask, request, jsonify
import requests
from bs4 import BeautifulSoup

app = Flask(__name__)

@app.route("/api/preview", methods=["POST"])
def preview():
    data = request.get_json()
    url = data.get("url")
    if not url:
        return jsonify({"error": "url is required"}), 400

    try:
        resp = requests.get(url, timeout=10)
        soup = BeautifulSoup(resp.text, "html.parser")

        title = soup.title.string if soup.title else ""
        desc_tag = soup.find("meta", attrs={"name": "description"})
        description = desc_tag["content"] if desc_tag else ""
        img_tag = soup.find("meta", attrs={"property": "og:image"})
        image = img_tag["content"] if img_tag else ""

        return jsonify({"title": title, "description": description, "image": image})
    except Exception as e:
        return jsonify({"error": str(e)}), 400

if __name__ == "__main__":
    app.run()

File: bench/testdata/samples/A10/PSB-A10-PREV-001/javascript/vulnerable.js
// PSB-A10-PREV-001: URL preview/unfurl feature
// CWE: CWE-918
// Expected: GTSS-SSRF-001
const express = require("express");
const axios = require("axios");
const cheerio = require("cheerio");
const app = express();
app.use(express.json());

app.post("/api/preview", async (req, res) => {
  const { url } = req.body;
  if (!url) {
    return res.status(400).json({ error: "url is required" });
  }

  try {
    const resp = await axios.get(url, { timeout: 10000 });
    const $ = cheerio.load(resp.data);

    const title = $("title").text() || "";
    const description = $('meta[name="description"]').attr("content") || "";
    const image = $('meta[property="og:image"]').attr("content") || "";

    res.json({ title, description, image });
  } catch (err) {
    res.status(400).json({ error: "failed to fetch URL" });
  }
});

module.exports = app;
